test/cqlpy/test_filtering.py

```diff
@@ -152,6 +152,30 @@ def test_filtering_with_in_relation(cql, test_keyspace, cassandra_bug):
         res = cql.execute(f"select * from {table} where v in (5,7) ALLOW FILTERING")
         assert set(res) == set([(2,3,4,5), (4,5,6,7)])
 
+# Test that NOT IN restrictions are supported with filtering and return the
+# correct results.
+# We mark this test "cassandra_bug" because Cassandra could support this
+# feature but doesn't yet: It reports "IN predicates on non-primary-key
+# columns (v) is not yet supported" when v is a regular column, or "IN
+# restrictions are not supported when the query involves filtering" on
+# partition-key columns p1 or p2. By the way, it does support IN restrictions
+# on a clustering-key column.
+def test_filtering_with_in_relation(cql, test_keyspace, cassandra_bug):
+    schema = 'p1 int, p2 int, c int, v int, primary key ((p1, p2),c)'
+    with new_test_table(cql, test_keyspace, schema) as table:
+        cql.execute(f"INSERT INTO {table} (p1, p2, c, v) VALUES (1, 2, 3, 4)")
+        cql.execute(f"INSERT INTO {table} (p1, p2, c, v) VALUES (2, 3, 4, 5)")
+        cql.execute(f"INSERT INTO {table} (p1, p2, c, v) VALUES (3, 4, 5, 6)")
+        cql.execute(f"INSERT INTO {table} (p1, p2, c, v) VALUES (4, 5, 6, 7)")
+        res = cql.execute(f"select * from {table} where p1 NOT IN (2,4) ALLOW FILTERING")
```

Comment:
Interesting - before I thought about this issue thoroughly, I thought that these request would not need "ALLOW FILTERING", but now I changed my mind: Consider that you have a terabyte of data and all of it has p1=2 - then a query of "p1 NOT IN (2,4)" will need to read this entire terabyte and not return anything.

So please add a test that the same queries without ALLOW FILTERING don't work.  We also have test_allow_filtering.py that has a somewhat more "clever" framework to check ALLOW FILTERING that you can consider using (but don't have to).

I think, but it should perhaps be discussed, that "ALLOW FILTERING" should not be mandatory in these requests. The concept of "ALLOW FILTERING" is about a situation

Reply:
> Interesting - before I thought about this issue thoroughly, I thought that these request would not need "ALLOW FILTERING", but now I changed my mind: Consider that you have a terabyte of data and all of it has p1=2 - then a query of "p1 NOT IN (2,4)" will need to read this entire terabyte and not return anything.
> 
> So please add a test that the same queries without ALLOW FILTERING don't work. 

Sure.

> We also have test_allow_filtering.py that has a somewhat more "clever" framework to check ALLOW FILTERING that you can consider using (but don't have to).
> 
> I think, but it should perhaps be discussed, that "ALLOW FILTERING" should not be mandatory in these requests. The concept of "ALLOW FILTERING" is about a situation

It is indeed.

Reply:
The last paragraph I wrote makes no sense. I think I started to write something and then in the middle switch to write the first paragraphs - and the incomplete last paragraph remained ;-)